File: services/core-api/src/handlers/devices.rs
use axum::{
    extract::{State, Query},
    Json,
    headers::{Authorization, Bearer},
};
use sqlx::PgPool;
use uuid::Uuid;

use crate::error::AppError;
use crate::models::{BindDeviceRequest, BindDeviceResponse, Device};
use crate::AppState;

pub async fn bind_device(
    State((pool, _): State<(sqlx::PgPool, AppState)>),
    headers: Authorization<Bearer>,
    Json(payload): Json<BindDeviceRequest>,
) -> Result<Json<BindDeviceResponse>, AppError> {
    let token = headers.token();
    let user = get_user_by_token(&pool, token).await?;

    let device: Device = sqlx::query_as(
        r#"
        INSERT INTO devices (user_id, device_id, device_type, device_name, os_version, app_version)
        VALUES ($1, $2, $3, $4, $5, $6)
        ON CONFLICT (device_id) DO UPDATE SET
            user_id = EXCLUDED.user_id,
            device_name = EXCLUDED.device_name,
            os_version = EXCLUDED.os_version,
            app_version = EXCLUDED.app_version,
            last_sync_at = NOW(),
            is_active = true
        RETURNING id, user_id, device_id, device_type, device_name, os_version, app_version,
                  last_sync_at, is_active, created_at
        "#
    )
    .bind(user.id)
    .bind(&payload.device_id)
    .bind(&payload.device_type)
    .bind(&payload.device_name)
    .bind(&payload.os_version)
    .bind(&payload.app_version)
    .fetch_one(&pool)
    .await?;

    Ok(Json(BindDeviceResponse { device }))
}

pub async fn list_devices(
    State((pool, _): State<(sqlx::PgPool, AppState)>),
    headers: Authorization<Bearer>,
) -> Result<Json<Vec<Device>>, AppError> {
    let token = headers.token();
    let user = get_user_by_token(&pool, token).await?;

    let devices: Vec<Device> = sqlx::query_as(
        r#"
        SELECT id, user_id, device_id, device_type, device_name, os_version, app_version,
               last_sync_at, is_active, created_at
        FROM devices
        WHERE user_id = $1 AND is_active = true
        ORDER BY last_sync_at DESC NULLS LAST
        "#
    )
    .bind(user.id)
    .fetch_all(&pool)
    .await?;

    Ok(Json(devices))
}

pub async fn get_user_by_token(pool: &PgPool, token: &str) -> Result<User, AppError> {
    let user = sqlx::query_as(
        r#"
        SELECT id, anonymous_token, nickname, avatar_url, campus, enrollment_year,
               risk_level, state_vector_id, total_interactions, created_at, updated_at, deleted_at
        FROM users
        WHERE anonymous_token = $1 AND deleted_at IS NULL
        "#
    )
    .bind(token)
    .fetch_optional(pool)
    .await?
    .ok_or_else(|| AppError::Unauthorized)?;

    Ok(user)
}

use crate::models::User;

File: services/ts-api/src/routes/devices.ts
import { Router, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { execute, query } from '../db';
import { AuthenticatedRequest, authMiddleware } from '../middleware/auth';
import { Device } from '../models';

const router = Router();

router.use(authMiddleware);

router.post('/bind', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { device_id, device_type, device_name } = req.body;
    const userId = req.user!.id;

    const existing = await query<Device>(
      'SELECT * FROM devices WHERE user_id = $1 AND device_id = $2',
      [userId, device_id]
    );

    if (existing.length > 0) {
      await execute(
        `UPDATE devices SET device_type = $1, device_name = $2, last_sync_at = $3, is_active = 1
         WHERE user_id = $4 AND device_id = $5`,
        [device_type, device_name, new Date().toISOString(), userId, device_id]
      );
    } else {
      const deviceId = uuidv4();
      await execute(
        `INSERT INTO devices (id, user_id, device_id, device_type, device_name, is_active, created_at)
         VALUES ($1, $2, $3, $4, $5, 1, $6)`,
        [deviceId, userId, device_id, device_type || null, device_name || null, new Date().toISOString()]
      );
    }

    const device = await query<Device>(
      'SELECT * FROM devices WHERE user_id = $1 AND device_id = $2',
      [userId, device_id]
    );

    res.status(201).json({ device: device[0] });
  } catch (error) {
    console.error('Bind device error:', error);
    res.status(500).json({ error: 'Failed to bind device' });
  }
});

router.get('/', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const userId = req.user!.id;
    const devices = await query<Device>(
      'SELECT * FROM devices WHERE user_id = $1 AND is_active = 1 ORDER BY last_sync_at DESC',
      [userId]
    );
    res.json({ devices });
  } catch (error) {
    console.error('List devices error:', error);
    res.status(500).json({ error: 'Failed to list devices' });
  }
});

export default router;
